aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -338,39 +322,38 @@ void dirichlet_scalar_cuda_kernel(
 
 } // namespace
 
-namespace at {
-namespace detail {
-
-Generator* getDefaultCUDAGenerator() {
-  globalContext().lazyInitCUDA();
-  return globalContext().generator_registry[static_cast<int>(DeviceType::CUDA)].get(); 
-}
-
-} // namespace detail
-
-namespace native {
+namespace at { namespace native {
 
-Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
+Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen_) {
+  auto gen = get_generator_or_default<CUDAGenerator>(gen_, detail::getCUDAHooks().getDefaultCUDAGenerator());
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(gen->mutex_);
```

Comment:
Aren't you taking the lock out for too large a scope here? You only need it to protect `gen->philox_engine_inputs`, but if you put the lock here, you'll take out the lock for the entirety of the kernel. (I suspect other occurrences are also falling prey to this problem, I won't comment on them.)